File: plugins/GameThrive/proj.android/src/org/cocos2dx/plugin/GameThrivePush.java
/**
 * Copyright 2014 GameThrive
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.cocos2dx.plugin;

import org.json.JSONObject;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;

import com.gamethrive.GameThrive;
import com.gamethrive.GameThrive.GetTagsHandler;
import com.gamethrive.GameThrive.IdsAvailableHandler;
import com.gamethrive.NotificationOpenedHandler;

public class GameThrivePush implements InterfacePush, PluginListener {
	private Context mContext;
	private GameThrive gameThrive;
	private static GameThrivePush mInstance;
	
	public GameThrivePush(Context context) {
		mContext = context;
		mInstance = this;
	}
	
	public void init(String googleProjectNumber, String appId) {
		gameThrive = new GameThrive((Activity)mContext, googleProjectNumber, appId, new NotificationOpenedHandler() {
			public void notificationOpened(String message, JSONObject additionalData, boolean isActive) {
				String curAdditionalData = null;
				if (additionalData != null)
					curAdditionalData = additionalData.toString();
				PushWrapper.onDidReceiveRemoteNotification(mInstance, message, curAdditionalData, isActive);
			}
		});
		
		PluginWrapper.addListener(this);
	}
	
	public void sendTag(String key, String value) {
		if (gameThrive != null)
			gameThrive.sendTag(key, value);
	}
	
	public void deleteTag(String key) {
		if (gameThrive != null)
			gameThrive.deleteTag(key);
	}
	
	public void getTags() {
		if (gameThrive != null) {
			gameThrive.getTags(new GetTagsHandler() {
				@Override
				public void tagsAvailable(JSONObject tags) {
					try {
						PushWrapper.onReceivedTags(mInstance, tags.toString());
					} catch (Throwable t) {
						t.printStackTrace();
					}
				}
			});
		}
	}
	
	public void getIds() {
		if (gameThrive != null) {
			gameThrive.idsAvailable(new IdsAvailableHandler() {
				@Override
				public void idsAvailable(String playerId, String registrationId) {
					try {
						PushWrapper.onReceivedIds(mInstance, playerId, registrationId);
					} catch (Throwable t) {
						t.printStackTrace();
					}
				}
			});
		}
	}

	@Override
	public void onResume() {
		if (gameThrive != null)
			gameThrive.onResumed();
	}

	@Override
	public void onPause() {
		if (gameThrive != null)
			gameThrive.onPaused();
	}

	@Override
	public void onDestroy() {
	}

	@Override
	public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
		return false;
	}
}

File: plugins/GameThrive/proj.ios/GameThrivePush/GameThrivePush.m
/**
 * Copyright 2014 GameThrive
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <UIKit/UIKit.h>
#import <objc/runtime.h>

#import "PushWrapper.h"

#import <GameThrive/GameThrive.h>
#import "GameThrivePush.h"


GameThrivePush* gameThrivePush;
GameThrive* gameThrive;

NSString* launchMessage;
NSDictionary* launchAdditionalData;
BOOL launchIsActive;

BOOL initCalled = false;


NSString* dictionaryToJsonString(NSDictionary* dictionaryToConvert) {
    if (!dictionaryToConvert)
        return nil;
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:dictionaryToConvert options:0 error:nil];
    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
}

void processNotificationOpened() {
    [PushWrapper onDidReceiveRemoteNotification:gameThrivePush message:launchMessage additionalData:dictionaryToJsonString(launchAdditionalData) isActive:launchIsActive];
}

void initGameThriveObject(NSDictionary* launchOptions, NSString* appId, BOOL autoRegister) {
    if (gameThrive == nil) {
        gameThrive = [[GameThrive alloc] initWithLaunchOptions:launchOptions appId:appId handleNotification:^(NSString* message, NSDictionary* additionalData, BOOL isActive) {
            launchMessage = message;
            launchAdditionalData = additionalData;
            launchIsActive = isActive;
            
            if (initCalled)
                processNotificationOpened();
        } autoRegister:autoRegister];
    }
}

@implementation UIApplication(GameThriveCocos2dxPush)

static void injectSelector(Class newClass, SEL newSel, Class addToClass, SEL makeLikeSel) {
    Method newMeth = class_getInstanceMethod(newClass, newSel);
    IMP imp = method_getImplementation(newMeth);
    const char* methodTypeEncoding = method_getTypeEncoding(newMeth);
    
    BOOL successful = class_addMethod(addToClass, makeLikeSel, imp, methodTypeEncoding);
    if (!successful) {
        class_addMethod(addToClass, newSel, imp, methodTypeEncoding);
        newMeth = class_getInstanceMethod(addToClass, newSel);
        
        Method orgMeth = class_getInstanceMethod(addToClass, makeLikeSel);
        
        method_exchangeImplementations(orgMeth, newMeth);
    }
}

+ (void)load {
    method_exchangeImplementations(class_getInstanceMethod(self, @selector(setDelegate:)), class_getInstanceMethod(self, @selector(setGameThriveCocos2dxDelegate:)));
}

static Class delegateClass = nil;

- (void) setGameThriveCocos2dxDelegate:(id<UIApplicationDelegate>)delegate {
    if(delegateClass != nil)
        return;
    delegateClass = [delegate class];
    
    injectSelector(self.class, @selector(gameThriveApplication:didFinishLaunchingWithOptions:),
                   delegateClass, @selector(application:didFinishLaunchingWithOptions:));
    [self setGameThriveCocos2dxDelegate:delegate];
}

- (BOOL)gameThriveApplication:(UIApplication*)application didFinishLaunchingWithOptions:(NSDictionary*)launchOptions {
    if ([launchOptions objectForKey:UIApplicationLaunchOptionsRemoteNotificationKey] != nil)
        initGameThriveObject(launchOptions, nil, true);
    
    if ([self respondsToSelector:@selector(gameThriveApplication:didFinishLaunchingWithOptions:)])
        return [self gameThriveApplication:application didFinishLaunchingWithOptions:launchOptions];
    return YES;
}

@end

@implementation GameThrivePush

- (void) initWithAppId:(NSString*)appId autoRegister:(BOOL)autoRegister {
    gameThrivePush = self;
    initCalled = true;
    initGameThriveObject(nil, appId, autoRegister);
    
    if (launchMessage)
        processNotificationOpened();
}

- (void) registerForPushNotifications {
    [gameThrive registerForPushNotifications];
}

- (void) sendTag:(NSString*)key value:(NSString*)value {
    [gameThrive sendTag:key value:value];
}

- (void) deleteTag:(NSString*)key {
    [gameThrive deleteTag:key];
}

- (void) getTags {
    [gameThrive getTags:^(NSDictionary* result) {
        [PushWrapper onReceivedTags:self  tags:dictionaryToJsonString(result)];
    }];
}

- (void) getIds {
    [gameThrive IdsAvailable:^(NSString* playerId, NSString* pushToken) {
        [PushWrapper onReceivedIds:self userid:playerId pushToken:pushToken];
    }];
}

@end
